### PR TITLE
Interfaces Support

### PR DESCRIPTION
**Pull request by haiodo**
Introduce Interfaces into model.

No search for interfaces, but it is possible to use Hierarchy descendants to find all classes.

Signed-off-by: Andrey Sobolev [haiodo@gmail.com](mailto:haiodo@gmail.com)

### CODE REVIEW

File: models/recruit/src/index.ts [hunk 1]
```diff
@@ -80,6 +76,10 @@ export class TApplicant extends TDocWithState implements Applicant {
 
   @Prop(TypeString(), 'Assigned recruiter' as IntlString)
   employee!: Ref<Employee>
+
+  // We need this two to make typescript happy.
+  declare state: TDocWithState['state']
```
Comment (haiodo) [inline models/recruit/src/index.ts:81]: Probable we could somehow exclude them for typescript with types, but I've not found an easy way.

File: packages/core/src/classes.ts [hunk 2]
```diff
@@ -118,12 +119,22 @@ export interface Classifier extends Doc, UXObject {
...
 
+/**
+ * @public
+ */
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+export interface Interface<T extends Obj> extends Classifier {
+  extends?: Ref<Interface<Obj>>[]
+  shortLabel?: IntlString
+}
+
```
Comment (aplatoff) [inline packages/core/src/classes.ts]: `Interface<T extends Doc>` is more correct. What is `shortLabel`? :) If we really need it in `Class` also why not move it to `Classifier`?

File: packages/core/src/hierarchy.ts [hunk 3]
```diff
@@ -96,10 +113,9 @@ export class Hierarchy {
   }
 
   private txCreateDoc (tx: TxCreateDoc<Doc>): void {
-    if (tx.objectClass === core.class.Class) {
-      const createTx = tx as TxCreateDoc<Class<Obj>>
-      const _id = createTx.objectId
-      this.classes.set(_id, TxProcessor.createDoc2Doc(createTx))
+    if ([core.class.Class, core.class.Interface].includes(tx.objectClass)) {
```
Comment (aplatoff) [inline packages/core/src/hierarchy.ts]: I believe this is much slower than ```tx.objectClass === core.class.Class || tx.objectClass ===core.class.Interface```

File: packages/model/src/dsl.ts [hunk 4]
```diff
@@ -105,17 +106,34 @@ export function Index (kind: IndexKind) {
...
+    txes.implements = _implements
     txes.domain = domain
     txes.kind = ClassifierKind.CLASS
   }
 }
 
+/**
+ * @public
+ */
+export function ModelInterface<T extends Obj> (
```
Comment (aplatoff) [inline packages/model/src/dsl.ts]: Let's change `ModelInterface` to `Implements`? for better model look

**Review by aplatoff — CHANGES_REQUESTED** [review]
Look nice!

File: packages/core/src/classes.ts (hunk 2 above)
Comment (haiodo) [reply packages/core/src/classes.ts]: Not really needed for now, removed.

File: packages/model/src/dsl.ts (hunk 4 above)
Comment (haiodo) [reply packages/model/src/dsl.ts]: Done.

File: packages/core/src/hierarchy.ts (hunk 3 above)
Comment (haiodo) [reply packages/core/src/hierarchy.ts]: Fixed.

**Review by aplatoff — APPROVED** [review]

**Pull request merged**